controller/internal/windows/iptables_windows.go

```diff
@@ -0,0 +1,337 @@
+// +build windows
+
+package windows
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/DavidGamba/go-getoptions"
+	"go.aporeto.io/trireme-lib/controller/internal/windows/frontman"
+	"go.aporeto.io/trireme-lib/controller/pkg/packet"
+)
+
+// structure representing result of parsed --match-set
+type WindowsRuleMatchSet struct {
+	MatchSetName    string
+	MatchSetNegate  bool
+	MatchSetDstIp   bool
+	MatchSetDstPort bool
+	MatchSetSrcIp   bool
+	MatchSetSrcPort bool
+}
+
+// structure representing parsed port range
+type WindowsRulePortRange struct {
+	PortStart int
+	PortEnd   int
+}
+
+// structure representing result of parsed iptables rule
+type WindowsRuleSpec struct {
+	Protocol         int
+	Action           int // FilterAction (allow, drop, nfq, proxy)
+	ProxyPort        int
+	Mark             int
+	Log              bool
+	LogPrefix        string
+	GroupId          int
+	MatchSrcPort     []*WindowsRulePortRange
+	MatchDstPort     []*WindowsRulePortRange
+	MatchBytes       []byte
+	MatchBytesOffset int
+	MatchSet         []*WindowsRuleMatchSet
+}
+
+// converts a WindowsRuleSpec back into a string for an iptables rule
+func MakeRuleSpecText(winRuleSpec *WindowsRuleSpec) string {
+	rulespec := ""
+	if winRuleSpec.Protocol > 0 && winRuleSpec.Protocol < math.MaxUint8 {
+		rulespec += fmt.Sprintf("-p %d ", winRuleSpec.Protocol)
+	}
+	if len(winRuleSpec.MatchBytes) > 0 {
+		rulespec += fmt.Sprintf("-m string --string %s --offset %d ", string(winRuleSpec.MatchBytes), winRuleSpec.MatchBytesOffset)
+	}
+	if len(winRuleSpec.MatchSrcPort) > 0 {
+		rulespec += "--sports "
+		for i, pr := range winRuleSpec.MatchSrcPort {
+			rulespec += strconv.Itoa(pr.PortStart)
+			if pr.PortStart != pr.PortEnd {
+				rulespec += fmt.Sprintf(":%d", pr.PortEnd)
+			}
+			if i+1 < len(winRuleSpec.MatchSrcPort) {
+				rulespec += ","
+			}
+		}
+		rulespec += " "
+	}
+	if len(winRuleSpec.MatchDstPort) > 0 {
+		rulespec += "--dports "
+		for i, pr := range winRuleSpec.MatchDstPort {
+			rulespec += strconv.Itoa(pr.PortStart)
+			if pr.PortStart != pr.PortEnd {
+				rulespec += fmt.Sprintf(":%d", pr.PortEnd)
+			}
+			if i+1 < len(winRuleSpec.MatchDstPort) {
+				rulespec += ","
+			}
+		}
+		rulespec += " "
+	}
+	if len(winRuleSpec.MatchSet) > 0 {
+		for _, ms := range winRuleSpec.MatchSet {
+			rulespec += "-m set "
+			if ms.MatchSetNegate {
+				rulespec += "! "
+			}
+			rulespec += fmt.Sprintf("--match-set %s ", ms.MatchSetName)
+			if ms.MatchSetSrcIp {
+				rulespec += "srcIP"
+				if ms.MatchSetSrcPort || ms.MatchSetDstPort {
+					rulespec += ","
+				}
+			} else if ms.MatchSetDstIp {
+				rulespec += "dstIP"
+				if ms.MatchSetSrcPort || ms.MatchSetDstPort {
+					rulespec += ","
+				}
+			}
+			if ms.MatchSetSrcPort {
```

Comment:
What happens if none matches?

Reply:
I had assumed that the passed in WinRuleSpec would have been created via a previous parse, which should guarantee that the rulespec is valid. But since this is a public function, I can't guarantee it, so I modified the code by adding a 'validate bool', which if true will do a final parse and check for error. This keeps the 'rulespec validity check' code in one place, too.